### PR TITLE
Move fos rest, jms serializer and hateoas on optional requirements

### PR DESCRIPTION
**Pull request by loic425**
| Q               | A
| --------------- | -----
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | mentioned in upgrade
| Deprecations?   | no
| Related tickets | 
| License         | MIT

It's time to get rid of these packages. Let's start by moving them on optional requirements.

- [x] move on dev dependencies
- [x] fix DI
- [x] FosRestPassTest
- [x] HateoasPassTest

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -1,3 +1,13 @@
+## UPGRADE FOR `1.12.x`
+
+### FROM `1.11.x` to `1.12.x`
+
+We move these following dependencies on optional requirements, so explicit them in your app requirements if you need them.
```
Comment (diimpp) [inline UPGRADE.md]: My take, maybe someone would like to take a look at this as well.

```suggestion
In preparation of removal, following dependencies were moved to optional requirements. If still in use by your app, require them explicitly in your `composer.json`.
```

File: src/Bundle/DependencyInjection/Compiler/FosRestPass.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ResourceBundle\DependencyInjection\Compiler;
+
+use FOS\RestBundle\FOSRestBundle;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+final class FosRestPass implements CompilerPassInterface
```
Comment (diimpp) [inline src/Bundle/DependencyInjection/Compiler/FosRestPass.php]: Would be cool to make more explicit name, like

```suggestion
final class UnregisterDefinitionsOnNonInstalledFOSRestBundlePass implements CompilerPassInterface
```

I'm not suggesting this particular name though, but stating intention in name (Unregister) and more full vendor name (FOSRestBundle) looks better to me.

File: src/Bundle/DependencyInjection/Compiler/HateoasPass.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ResourceBundle\DependencyInjection\Compiler;
+
+use Bazinga\Bundle\HateoasBundle\BazingaHateoasBundle;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+final class HateoasPass implements CompilerPassInterface
```
Comment (diimpp) [inline src/Bundle/DependencyInjection/Compiler/HateoasPass.php]:
```suggestion
final class UnregisterDefinitionsForMissingBazingaHateoasBundlePass implements CompilerPassInterface
```

File: src/Bundle/DependencyInjection/Compiler/FosRestPass.php (hunk 2 above)
Comment (loic425) [reply src/Bundle/DependencyInjection/Compiler/FosRestPass.php]: I agree and I've changed that with sth between previous name and yours.

File: src/Bundle/DependencyInjection/Compiler/HateoasPass.php (hunk 3 above)
Comment (loic425) [reply src/Bundle/DependencyInjection/Compiler/HateoasPass.php]: same

**Review by diimpp — APPROVED** [review]

**Review by NoResponseMate — APPROVED** [review]

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Loïc! :tada:

**Pull request merged**